### PR TITLE
EVG-6433 abort/dequeue failed commit queue versions

### PR DESCRIPTION
**Pull request by ablack12**
When a commit queue version has a task fail, we want to abort the version and dequeue the commit queue item so we don't hold up the queue.  

This means whenever a project has a task that doesn't succeed and the commit queue enabled, we'll have to get the patch. (Should this specifically be task status == failed?)

Let me know if I've made an invalid assumption about how versions/commit queues work.

### CODE REVIEW

File: model/task_lifecycle.go [hunk 1]
```diff
@@ -499,6 +501,38 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
...
+		return nil
+	}
+
+	// if task is part of a commit queue, dequeue and abort version
+	cq, err := commitqueue.FindOneId(projectId)
+	if err != nil {
+		return errors.Wrapf(err, "can't get commit queue for id '%s'", projectId)
+	}
+	issue := p.Id.Hex()
+	if p.IsGithubPRPatch() {
```
Comment (ybrill) [inline model/task_lifecycle.go]: in the case of a PR commit queue do we also want to send a status to GitHub?

File: model/task_lifecycle_test.go [hunk 2]
```diff
@@ -1175,6 +1177,101 @@ func TestAbortTask(t *testing.T) {
...
+		}
+		if thisTask.Status == evergreen.TaskUndispatched {
+			assert.False(t, thisTask.Activated)
+		}
+	}
+	assert.Equal(t, 2, aborted)
+	assert.Equal(t, 1, finished)
+	p, err = patch.FindOne(patch.ByVersion("my-version"))
+	assert.NoError(t, err)
+	assert.NotNil(t, p)
```
Comment (tychoish) [inline model/task_lifecycle_test.go]: might be more clear to break this case into named subtests

File: model/task_lifecycle.go [hunk 3]
```diff
@@ -499,6 +501,38 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
...
+
+	if p.Alias != evergreen.CommitQueueAlias {
+		return nil
+	}
+
+	// if task is part of a commit queue, dequeue and abort version
+	cq, err := commitqueue.FindOneId(projectId)
+	if err != nil {
+		return errors.Wrapf(err, "can't get commit queue for id '%s'", projectId)
+	}
```
Comment (tychoish) [inline model/task_lifecycle.go]: are we worried about cq=nil?

File: model/task_lifecycle.go [hunk 4]
```diff
@@ -499,6 +501,38 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
...
+	issue := p.Id.Hex()
+	if p.IsGithubPRPatch() {
+		issue = strconv.Itoa(p.GithubPatchData.PRNumber)
+	}
+	removed, err := cq.Remove(issue)
+	if err != nil {
+		return errors.Wrapf(err, "can't remove item '%s' from queue '%s'", versionId, projectId)
+	}
+	if !removed {
+		return errors.Errorf("Item not on queue")
```
[the remark this answers is not in the document]
Comment (tychoish) [reply model/task_lifecycle.go]: this should be either errors.New() or should include info about the version/porject/etc.

File: model/task_lifecycle.go [hunk 5]
```diff
@@ -499,6 +501,38 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
...
+	}
+	return errors.Wrapf(CancelPatch(p, requester), "Error aborting failed commit queue patch")
```
[the remark this answers is not in the document]
Comment (ablack12) [reply model/task_lifecycle.go]: Maybe the question should be if it's not on the commit queue should we still try to cancel the patch. I would argue no since it's probably done if it's not on the queue anymore.

Comment (ybrill) [inline model/task_lifecycle.go]: if the patch is removed from the queue we should log `event.LogCommitQueueConcludeTest(issue, status)` so the user is notified.

File: service/api_task.go [hunk 6]
```diff
@@ -194,6 +194,19 @@ func (as *APIServer) EndTask(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if projectRef.CommitQueue.Enabled && details.Status != evergreen.TaskSucceeded {
+		env := evergreen.GetEnvironment()
+		githubStatusSender, err := env.GetSender(evergreen.SenderGithubStatus)
```
Comment (john-m-liu) [inline service/api_task.go]: I would move this into TryDequeueAndAbortCommitQueueVersion since there's no need to get a sender in the API here

File: model/task_lifecycle.go [hunk 7]
```diff
@@ -3,11 +3,15 @@ package model
 import (
 	"fmt"
 	"sort"
+	"strconv"
 	"time"
 
+	"github.com/mongodb/grip/send"
+
```
Comment (tychoish) [inline model/task_lifecycle.go]: nit: blank line

File: model/task_lifecycle.go [hunk 8]
```diff
@@ -499,6 +501,62 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
...
+			grip.Debug(message.WrapError(err, message.Fields{
+				"message":      "error getting sender for github status",
+				"patch_id":     p.Id,
+				"project":      p.Project,
+				"pull_request": issue,
+			}))
+		} else {
+			status := message.GithubStatus{
+				Context:     commitqueue.Context,
+				Description: "removed from queue because task failed",
```
Comment (ybrill) [inline model/task_lifecycle.go]: Since this is what the user will see if their patch fails and we refuse to merge perhaps the message should simply be "merge test failed", as they would see now when the patch finishes (https://github.com/evergreen-ci/evergreen/blob/9713dee32db66be286d91fd254ef9d6bba2acbbe/model/commitqueue/github_pr_sender.go#L150)

File: model/task_lifecycle.go [hunk 9]
```diff
@@ -499,6 +501,62 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
...
+				State:       message.GithubStateFailure,
+				Ref:         p.GithubPatchData.HeadHash,
+				Repo:        p.GithubPatchData.HeadRepo,
+				Owner:       p.GithubPatchData.Author,
```
Comment (ybrill) [inline model/task_lifecycle.go]: I think the status will only make it to the PR with the owner/repo of the base repo (like for evergreen, the owner would be `evergreen-ci` instead of `ablack12`)

File: model/task_lifecycle.go [hunk 10]
```diff
@@ -499,6 +501,62 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
...
+				"patch_id":     p.Id,
+				"project":      p.Project,
+				"pull_request": issue,
+			}))
+		} else {
+			status := message.GithubStatus{
+				Context:     commitqueue.Context,
+				Description: "removed from queue because task failed",
+				State:       message.GithubStateFailure,
+				Ref:         p.GithubPatchData.HeadHash,
```
Comment (ybrill) [inline model/task_lifecycle.go]: I think this information is not currently set in commit queue patches
https://github.com/evergreen-ci/evergreen/blob/9713dee32db66be286d91fd254ef9d6bba2acbbe/model/patch/patch.go#L401

File: model/commitqueue/github_pr_sender.go [hunk 11]
```diff
@@ -164,6 +166,37 @@ func (s *githubPRLogger) sendPatchResult(msg *GithubMergePR, pr event.PRInfo) {
...
+	}
+	ghToken, err := env.Settings().GetGithubOauthToken()
+	if err != nil {
+		return errors.Wrap(err, "error getting github oauth token")
+	}
+
+	ctxWithCancel, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	pr, err := thirdparty.GetGithubPullRequest(ctxWithCancel, ghToken, owner, repo, prNumber)
```
Comment (ybrill) [inline model/commitqueue/github_pr_sender.go]: an alternative to getting the PR from GitHub could be to save the head SHA in the patch when we create it
https://github.com/evergreen-ci/evergreen/blob/72262b72b08104da687e13ce48fd7d76ff029be9/model/patch/patch.go#L401

File: model/commitqueue/github_pr_sender.go [hunk 12]
```diff
@@ -164,6 +166,37 @@ func (s *githubPRLogger) sendPatchResult(msg *GithubMergePR, pr event.PRInfo) {
 	s.statusSender.Send(c)
 }
 
+func SendFailedMessageToPR(owner, repo string, prNumber int) error {
```
[the remark this answers is not in the document]
Comment (ablack12) [reply model/commitqueue/github_pr_sender.go]: Do you know where it would make more sense? task_lifecycle doesn't seem better.

File: model/commitqueue/github_pr_sender.go (hunk 11 above)
Comment (ablack12) [reply model/commitqueue/github_pr_sender.go]: It sounds like the owner is wrong too though. It looks like we get the PR from github in a lot of places so I'm not convinced sending around more data in patches is better than getting the PR again.

Comment (ybrill) [reply model/commitqueue/github_pr_sender.go]: why is the owner not the same?
https://github.com/evergreen-ci/evergreen/blob/72262b72b08104da687e13ce48fd7d76ff029be9/model/patch/patch.go#L404

Comment (ybrill) [reply model/commitqueue/github_pr_sender.go]: I would argue the more information we fill in for a patch's `GithubPatchData` the less confusing the code will be (so you don't have to keep track of what's been set), but I don't have a strong opinion.
I would argue against getting the PR if we don't need to to avoid GitHub API limits.

Comment (ablack12) [reply model/commitqueue/github_pr_sender.go]: Okay I'm convinced haha

File: model/commitqueue/github_pr_sender.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ybrill) [reply model/commitqueue/github_pr_sender.go]: maybe `thirdparty/github.go`?

[the remark this answers is not in the document]
Comment (ablack12) [reply model/commitqueue/github_pr_sender.go]: Now that we aren't fetching the PR I think this is short enough to leave in the original function.

File: self-tests.yml [hunk 13]
```diff
@@ -484,14 +484,20 @@ tasks:
           target: revendor
       - func: run-make
         vars: { target: "test-thirdparty-docker" }
+  - name: added-test
```
Comment (ybrill) [inline self-tests.yml]: were these changes intentional?

File: model/task_lifecycle.go [hunk 14]
```diff
@@ -499,6 +501,69 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
...
+			if err := uiConfig.Get(env); err == nil {
+				url = fmt.Sprintf("%s/version/%s", uiConfig.Url, versionId)
+			}
+			status := message.GithubStatus{
+				Context:     commitqueue.Context,
+				Description: "merge test failed",
+				State:       message.GithubStateFailure,
+				Owner:       p.GithubPatchData.BaseOwner,
+				Repo:        p.GithubPatchData.BaseRepo,
+				Ref:         p.GithubPatchData.BaseHash,
```
Comment (ybrill) [inline model/task_lifecycle.go]: I think we need to send it to the head hash, similar to here
https://github.com/evergreen-ci/evergreen/blob/72262b72b08104da687e13ce48fd7d76ff029be9/units/commit_queue.go#L481

Comment (ablack12) [reply model/task_lifecycle.go]: Okay I misunderstood this at first but now I'm using HeadHash to store the Head.SHA which shouldn't conflict with other uses of it I believe?
